### ISSUE TITLE
Undefined pointer comparison in macro.c

### ISSUE THREAD

**Issue by DemiMarie** [rpm-software-management/rpm#1602]
GCC has `-fsanitize=pointer-compare`, which adds instrumentation to detect invalid pointer comparisons.  When built with this flag, and run with `ASAN_OPTIONS=detect_invalid_pointer_pairs=1`, virtually the entire testsuite fails due to [an undefined pointer comparison][1] in `rpmio/macro.c`.  This is in turn due to [`q` being initialized as one before the beginning of an allocated region][2].

Reproduction script below.  PR coming.

[1]: https://github.com/rpm-software-management/rpm/blob/e838c489dbbf37c0476c423a7555ad4672eb075d/rpmio/macro.c#L229
[2]: https://github.com/rpm-software-management/rpm/blob/e838c489dbbf37c0476c423a7555ad4672eb075d/rpmio/macro.c#L212

[build-rpm.gz](https://github.com/rpm-software-management/rpm/files/6210103/build-rpm.gz)

**Comment by pmatilai** [comment]
That is probably undefined behavior in C11 but AIUI not in C99, which rpm uses.
Perhaps I wasn't clear enough in my earlier comment today: as a rule of thumb, I'm *not* interested in undefined behavior findings, exactly because of it's totally undefined what standard it's checking.

**Comment by DemiMarie** [comment]
> That is probably undefined behavior in C11 but AIUI not in C99, which rpm uses.

It is undefined in C99; see http://www.open-std.org/jtc1/sc22/wg14/www/C99RationaleV5.10.pdf, §6.3.2.3:

> Regardless how an invalid pointer is created, any use of it yields undefined behavior.

So I still believe it should be fixed.

> Perhaps I wasn't clear enough in my earlier comment today: as a rule of thumb, I'm _not_ interested in undefined behavior findings, exactly because of it's totally undefined what standard it's checking. Unless it reveals a concrete bug of course.

This is undefined behavior in C99, and includes a reproducible test case.  I consider that to be a concrete bug.

**Comment by mlschroe** [comment]
Maybe I'm dense, but isn't q incremented in line 221 before the pointer comparison is done?

**Comment by DemiMarie** [comment]
It gets decremented again on line 227.

**Comment by mlschroe** [comment]
So the initialization is not the problem? And a fix would be do change the for loop to:
```
for (; nb && p <= q; p++) {
```

**Comment by DemiMarie** [comment]
> So the initialization is not the problem? And a fix would be do change the for loop to:
> 
> ```
> for (; nb && p <= q; p++) {
> ```

The initialization is already undefined behavior.

**Comment by mlschroe** [comment]
The `if (nb == 0` in line 253 is most likely wrong as it is true for empty lines. I guess it should read:
```
if ((nb && *q != '\\' && !bc && !pc && !xc) || *(q+1) == '\0') {
```
Panu, do you agree?

**Comment by mlschroe** [comment]
I also don't think rpm will work correctly on EBCDIC platforms.

**Comment by mlschroe** [comment]
Of course the initialization can be easily fixed by doing a
```
*q = 0;  /* terminate */
```
at the top of the do loop and
```
q++;    /* move forward */
```
at the bottom. That way, q can be initialized to buf. (It will be decremented to buf-1 for empty lines nevertheless.)

**Comment by mlschroe** [comment]
Regarding my comment about the wrong nb check:

The following macro definition works:
```
%hello %{expand:
hi
ho
}
```

But this produces an error:
```
%hello %{expand:
hi

ho
}
```

**Comment by mlschroe** [comment]
My suggested fix was wrong, it should be:
```
if (((!nb || *q != '\\') && !bc && !pc && !xc) || *(q+1) == '\0') {
```

Back to the pointer comparison: an even saner approach for this issue would be to let q point to the newline and not the char before the newline.

**Comment by pmatilai** [comment]
This is actually a *wonderful* example: just like compiler warnings, undefined behavior is best viewed as a canary that points to buggy code, rather than flaw in itself. Much, much damage to codebases around the world is done by "fixing" compiler warnings, and undefined behavior is not any different.

@mlschroe 's version fixes an ages old, very concrete bug (even an RFE) in the macro parser. It wasn't misbehaving because of the allegedly undefined behavior but because it was just a piece of bad old code, led to by the undefined behavior alert.

Fixed by #1606
